src/linear_algebra/orientation.lean

```diff
@@ -0,0 +1,376 @@
+lemma same_orientation_pos_smul_right (f : alternating_map R M R ι) {r : R} (h : 0 < r) :
+  same_orientation f (r • f) :=
+⟨r, 1, h, zero_lt_one, (one_smul _ _).symm⟩
+
+/-- A positive multiple of an alternating map has the same orientation as that map. -/
+lemma same_orientation_pos_smul_left (f : alternating_map R M R ι) {r : R} (h : 0 < r) :
+  same_orientation (r • f) f :=
+⟨1, r, zero_lt_one, h, one_smul _ _⟩
+
+end nontrivial
+
+variables {R M ι}
+
+/-- If two alternating maps have the same orientation, so do their negations. -/
+@[simp] lemma same_orientation_neg_iff (f g : alternating_map R M R ι) :
+  same_orientation (-f) (-g) ↔ same_orientation f g :=
+⟨λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj] at h⟩,
+ λ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₁, r₂, hr₁, hr₂, by rwa [smul_neg, smul_neg, neg_inj]⟩⟩
+
+end alternating_map
+
+variables (R M ι)
+
+/-- Nonzero alternating maps, as used to define orientations. -/
+@[reducible] def orientation_map := {f : alternating_map R M R ι // f ≠ 0}
+
+instance : has_coe_to_fun (orientation_map R M ι) (λ _, (ι → M) → R) :=
+⟨λ x, ⇑(x : alternating_map R M R ι)⟩
+
+namespace orientation_map
+
+variables {R M ι}
+
+/-- Negating a nonzero alternating map. -/
+instance : has_neg (orientation_map R M ι) :=
+⟨λ ⟨f, hf⟩, ⟨-f, by rw neg_ne_zero; exact hf⟩⟩
+
+/-- Negating a nonzero alternating map commutes with coercion to `alternating_map`. -/
+@[simp, norm_cast] lemma coe_neg (f : orientation_map R M ι) :
+  ((-f : orientation_map R M ι) : alternating_map R M R ι) = -(f : alternating_map R M R ι) :=
+by cases f; refl
+
+/-- Negating a nonzero alternating map twice produces the original map. -/
+@[simp] protected lemma neg_neg (f : orientation_map R M ι) : -(-f) = f :=
+by rw [subtype.ext_iff, coe_neg, coe_neg, neg_neg]
+
+variables (R M ι)
+
+variables [nontrivial R]
+
+/-- The setoid of the `same_orientation` relation for the subtype of nonzero alternating maps. -/
+instance same_orientation_setoid : setoid (orientation_map R M ι) :=
+{ r := λ f g, (f : alternating_map R M R ι).same_orientation g,
+  iseqv := equivalence.comap (alternating_map.equivalence_same_orientation R M ι) _ }
+
+variables {R M ι}
+
+/-- Equivalence of nonzero alternating maps, in terms of same_orientation. -/
+lemma equiv_iff_same_orientation (f g : orientation_map R M ι) :
+  f ≈ g ↔ (f : alternating_map R M R ι).same_orientation g :=
+iff.rfl
+
+/-- If two nonzero alternating maps are equivalent, so are their negations. -/
+@[simp] lemma equiv_neg_iff (f g : orientation_map R M ι) : -f ≈ -g ↔ f ≈ g :=
+by rw [equiv_iff_same_orientation, equiv_iff_same_orientation, coe_neg, coe_neg,
+       alternating_map.same_orientation_neg_iff]
+
+end orientation_map
+
+variables [nontrivial R]
+
+/-- An orientation of a module, intended to be used when `ι` is a `fintype` with the same
+cardinality as a basis. -/
+@[ext] class orientation :=
```

Comment:
It seems a bit weird to me to make this a class when every use in this file takes it as a regular argument

Reply:
I might be inclined to do something along the lines of:
```lean
def orientation := quotient (orientation_map.same_orientation_setoid R M ι)

class module.oriented :=
(positive_orientation : orientation)
```
and then just state all the lemmas about the former.
```

Reply:
I've now changed to separate `ray` (a quotient), `orientation` (an `abbreviation` for `ray` in the `alternating_map` case) and `module.oriented` (a class, as suggested).